### ISSUE TITLE
:w selects the whole file, thus losing caret placement.

### ISSUE THREAD

**Issue by LollipopFt** [helix-editor/helix#2752]
### Summary

https://user-images.githubusercontent.com/62802897/173198242-64d36aa5-4fb3-4b66-b15d-7ef8999015f3.mp4

as seen in the video above, when keying in `:w`, the entire file gets selected, and my caret always returns to the start when i exit it.

settings:
```toml
theme='onedark'
[keys.normal]
't'='move_line_down'
'n'='move_line_up'
's'='move_char_right'
'j'='find_till_char'
'k'='search_next'
'l'='select_regex'
[keys.insert]
','={'.'='normal_mode'}
[editor]
scrolloff=255
mouse=false
auto-completion=true
idle-timeout=0
completion-trigger-len=2
rulers=[80,100,120]
[editor.cursor-shape]
insert='bar'
[editor.whitespace]
render='all'
[editor.lsp]
display-messages=true
```

thank you!

### Reproduction Steps

 

I tried this:

1. `:w`

I expected this to happen:
no selections made.
Instead, this happened:
above video.

### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
none
```

</details>

### Platform

Windows

### Terminal Emulator

Windows Terminal Preview 1.14.1451.0

### Helix Version

helix 22.05 (27609f50)

**Comment by LollipopFt** [comment]
strangely enough, it only happens to one file:

https://user-images.githubusercontent.com/62802897/173198588-3916323b-26ad-4daf-8ec2-23159d3898af.mp4

any ideas what might be happening?

**Comment by sudormrfbin** [comment]
Could be due to line endings? (from #2404):

> I encountered this with rust-analyzer before. The line endings were different from what the LSP expected, and so formatting it would basically delete the entire document and send the new one with the correct line endings.

**Comment by RalfNorthman** [comment]
This happens to me too, after I upgraded to 22.05. Unfortunately I don't remember what my previous helix version was.

I've only seen it in rust-files when auto-format is on. I use rust-analyzer.

Running helix with the same configuration on a remote ubuntu machine works fine, so it seems to be windows-related.

### Platform
Windows 10

### Terminal Emulator
alacritty 0.10.0 (8a26dee)

### Helix Version
helix 22.05 (27609f50)

**Comment by LollipopFt** [comment]
@sudormrfbin possibly, i actually tried rewriting the entire file but it still caused the same problem. like i recreated it from scratch. so i have no idea what is happening...

**Comment by the-mikedavis** [comment]
I think that's consistent with the line-endings theory. IIRC, if you're on windows you'll use `crlf` line-endings by default. You could try using `:line-ending lf` before formatting to see if it's just the line-endings.

**Comment by RalfNorthman** [comment]
Yeah, when using `:line-ending lf` auto-format on `:w` works as expected!

Thanks!

**Comment by LollipopFt** [comment]
yep, can confirm this works. why does it only happen on `crlf`?
additional things: is there a way to see what line ending a file has like on windows notepad?
windows:
![image](https://user-images.githubusercontent.com/62802897/173329815-d2aef2a9-2521-4e50-a0b7-0b8490351e98.png)

**Comment by kirawi** [comment]
Perhaps it would also be possible to ask for rust-analyzer to be smarter about files with different line endings? I also wonder if other text editors have this issue with CRLF, and how they retain cursor position (if they do).

**Comment by farwyler** [comment]
For reference, [this](https://github.com/rust-lang/rust-analyzer/pull/7174/files) is the behaviour @the-mikedavis mentioned.
rust-analyzer sends the whole document if line-endings changed during formatting. 

But that would mean we only get this behaviour once. Currently it happens again, even after rust-analyzer changed the line endings. Presumably because helix changes the line endings right back?!

Looks like Document::detect_indent_and_line_ending is not called in this case. Maybe it should be called in general when LS sends a whole doc? Just like when the document is reloaded from disk?

**Comment by the-mikedavis** [comment]
Ah good catch @farwyler! Yes let's call `detect_indent_and_line_ending` when the LSP sends the whole document. Would you like to make a PR?

**Comment by farwyler** [comment]
@the-mikedavis will do

**Comment by farwyler** [comment]
Additionally...

i found a bug in rust-analyzer as well. When it sends the whole document because it wants to change the line endings, the replacement document actually uses the old line endings. This results in helix always receiving the whole document, because the endings never actually change.

I have created an issue here
https://github.com/rust-lang/rust-analyzer/issues/12540

**Comment by sudormrfbin** [comment]
Closed by #2778.
